docs/source/guide/events.rst

```diff
@@ -0,0 +1,445 @@
+Extensibility Guide
+===================
+
+All of Boto3's resource and client classes are generated at runtime.
+This means that you cannot directly inherit and then extend the
+functionality of these classes because they do not exist until the
+program actually starts running.
+
+
+However it is still possible to extend the functionality of classes through
+Boto3's event system.
+
+
+An Introduction to the Event System
+-----------------------------------
+
+Boto3's event system allows users to register a function to
+a specific event. Then once the running program reaches a line that
+emits that specific event, Boto3 will call every function
+registered to the event in the order in which they were registered.
+When Boto3 calls each of these registered functions,
+it will call each of them with a specific set of
+keyword arguments that are associated with that event.
+Then once the registered function
+is called, the function may modify the keyword arguments passed to that
+function or return a value.
+Here is an example of how the event system works::
+
+    import boto3
+
+    s3 = boto3.client('s3')
+
+    # Access the event system on the S3 client
+    event_system = s3.meta.events
+
+    # Create a function 
+    def my_function(my_list, **kwargs):
+        my_list.append('Called registered function')
+
+    # Register the function to an event
+    event_system.register('myevent', my_function)
+
+    orig_list = ['Before the event']
+
+    # Emit the event with the keyword argument my_list
+    event_system.emit('myevent', my_list=orig_list)
+    
+    orig_list.append('After the event')
+
+    print(orig_list)
+
+Here is the output from running this example::
+
+   ['Before the event', 'Called registered function', 'After the event']
+
+Here are the takeaways from this example:
+
+* All clients have their own event system that you can use to fire events
+  and register functions. You can access the event system through the
+  ``meta.events`` attribute on the client.
+* All functions registered to the event system must have ``**kwargs`` in
+  the function signature. This is because emitting an event can have any
+  number of keyword arguments emitted along side it, and so if your
+  function is called without ``**kwargs``, its signature will have to
```

Comment:
It's mostly for future compatibility as well.  It gives us the option of adding kwargs in the future without breaking existing handlers.